File: Data Structures & Algorithms/count-number-of-islands/submission-0.js
class Solution {
    /**
     * @param {character[][]} grid
     * @return {number}
     */
    numIslands(grid) {
        // Iterate through each cell of the grid
        // If a cell is '1' then it is land, and we can increment an island count.
        // We can use DFS to traverse this grid, marking all connected land cells as visited.

        let ROWS = grid.length;
        let COLS = grid[0].length;
        const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

        let islandCount = 0;

        const dfs = (row, col) => {
            if (row < 0 || row >= ROWS || col < 0 || col >= COLS || grid[row][col] === '0') {
                return;
            }

            grid[row][col] = '0';
            for (const [dr, dc] of directions) {
                dfs(row + dr, col + dc);
            }
        };
        
        for (let i = 0; i < ROWS; i++) {
            for (let j = 0; j < COLS; j++) {
                if (grid[i][j] === '1') {
                    islandCount++;
                    dfs(i, j);
                }
            }
        }
        return islandCount;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-0.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        cols = len(board[-1])
        
        def dfs(r, c, k, visited):
            # base case: if we've matched all characters of the word
            if k == len(word): return True

            if r < 0 or r >= rows or c < 0 or c >= cols or (r, c) in visited or board[r][c] != word[k]:
                return False

            visited.add((r, c))

            found = (dfs(r + 1, c, k + 1, visited) or
                    dfs(r - 1, c, k + 1, visited) or
                    dfs(r, c + 1, k + 1, visited) or
                    dfs(r, c - 1, k + 1, visited)
                    )
            visited.remove((r, c))

            return found
        
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == word[0]:
                    if dfs(r, c, 0, set()):
                        return True
        
        return False
